lease comparison in txn bug

### Bug report criteria

- [x] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [x] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [x] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [x] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

When I try to execute a lease comparison in a transaction (txn), etcdctl returns the error **"panic: bad value"** in goroutine. 

### What did you expect to happen?

Initially, the database is empty. I put the key value _{foo2: bar2}_ in the database, so I expected (replace \<hostname\> with your hostname):
```console
$ etcdctl --endpoints=<hostname>:2379 txn <<<'lease("foo1") > "0"

get foo1

get foo2

'
FAILURE

foo2
bar2
```

### How can we reproduce it (as minimally and precisely as possible)?

I started database with command (replace \<hostname\> with your hostname):
```console
$ etcd --listen-client-urls http://<hostname>:2379 --advertise-client-urls http://<hostname>:2379 --log-level 'debug'
```
In fact, you don't even need to run the database, because there is an error on the client.
Then I run:
```console
$ etcdctl --endpoints=<hostname>:2379 txn <<<'lease("foo1") > "0"

get foo1

get foo2

'
```



### Anything else we need to know?

I found a similar issue https://github.com/etcd-io/etcd/issues/14845 from three years ago, but it was closed, because an unsupported version was used in the bug report.

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
etcd Version: 3.7.0-alpha.0
Git SHA: 438ed9693
Go Version: go1.25.1
Go OS/Arch: linux/amd64

$ etcdctl version
etcdctl version: 3.7.0-alpha.0
API version: 3.7
```

</details>


### Etcd configuration (command line flags or environment variables)

_No response_

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

_No response_

### Relevant log output

```Shell
$ etcdctl --endpoints=<hostname>:2379 txn <<<'lease("foo1") > "0"               

get foo1

get foo2

'
panic: bad value

goroutine 1 [running]:
go.etcd.io/etcd/client/v3.mustInt64(...)
        go.etcd.io/etcd/client/v3/compare.go:132
go.etcd.io/etcd/client/v3.mustInt64orLeaseID(...)
        go.etcd.io/etcd/client/v3/compare.go:141
go.etcd.io/etcd/client/v3.Compare({0x1, 0x4, {0x0, 0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0, 0x0}, ...}, ...)
        go.etcd.io/etcd/client/v3/compare.go:66 +0x36c
go.etcd.io/etcd/etcdctl/v3/ctlv3/command.ParseCompare({0xc000154828, 0x13})
        go.etcd.io/etcd/etcdctl/v3/ctlv3/command/txn_command.go:225 +0x507
go.etcd.io/etcd/etcdctl/v3/ctlv3/command.readCompares(0xc000111c18)
        go.etcd.io/etcd/etcdctl/v3/ctlv3/command/txn_command.go:112 +0xc5
go.etcd.io/etcd/etcdctl/v3/ctlv3/command.txnCommandFunc(0xc0001ed808, {0xc0002f2dd0?, 0x4?, 0xd964e7?})
        go.etcd.io/etcd/etcdctl/v3/ctlv3/command/txn_command.go:79 +0x1e8
github.com/spf13/cobra.(*Command).execute(0xc0001ed808, {0xc0002f2db0, 0x1, 0x1})
        github.com/spf13/cobra@v1.10.1/command.go:1019 +0xae7
github.com/spf13/cobra.(*Command).ExecuteC(0x153bbc0)
        github.com/spf13/cobra@v1.10.1/command.go:1148 +0x465
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.10.1/command.go:1071
go.etcd.io/etcd/etcdctl/v3/ctlv3.Start(...)
        go.etcd.io/etcd/etcdctl/v3/ctlv3/ctl.go:121
go.etcd.io/etcd/etcdctl/v3/ctlv3.MustStart()
        go.etcd.io/etcd/etcdctl/v3/ctlv3/ctl.go:125 +0x1a
main.main()
        go.etcd.io/etcd/etcdctl/v3/main.go:23 +0xf
```

--- a/client/v3/compare.go
+++ b/client/v3/compare.go
@@ -15,6 +15,8 @@
 package clientv3
 
 import (
+	"fmt"
+
 	"google.golang.org/protobuf/proto"
 
 	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
@@ -172,7 +174,7 @@ func mustInt64(val any) int64 {
 	if v, ok := val.(int); ok {
 		return int64(v)
 	}
-	panic("bad value")
+	panic(fmt.Sprintf("bad value %v of type %T", val, val))
 }
 
 // mustInt64orLeaseID panics if val isn't a LeaseID, int or int64. It returns an

--- a/etcdctl/ctlv3/command/txn_command.go
+++ b/etcdctl/ctlv3/command/txn_command.go
@@ -23,7 +23,6 @@ import (
 	"strings"
 
 	"github.com/spf13/cobra"
-
 	clientv3 "go.etcd.io/etcd/client/v3"
 	"go.etcd.io/etcd/pkg/v3/cobrautl"
 )
@@ -221,7 +220,9 @@ func ParseCompare(line string) (*clientv3.Cmp, error) {
 	case "val", "value":
 		cmp = clientv3.Compare(clientv3.Value(key), op, val)
 	case "lease":
-		cmp = clientv3.Compare(clientv3.LeaseValue(key), op, val)
+		if v, err = strconv.ParseInt(val, 16, 64); err == nil {
+			cmp = clientv3.Compare(clientv3.LeaseValue(key), op, v)
+		}
 	default:
 		return nil, fmt.Errorf("malformed comparison: %s (unknown target %s)", line, target)
 	}

--- a/etcdctl/ctlv3/command/txn_command.go
+++ b/etcdctl/ctlv3/command/txn_command.go
@@ -23,6 +23,7 @@ import (
 	"strings"
 
 	"github.com/spf13/cobra"
+
 	clientv3 "go.etcd.io/etcd/client/v3"
 	"go.etcd.io/etcd/pkg/v3/cobrautl"
 )

--- a/etcdctl/ctlv3/command/txn_command_test.go
+++ b/etcdctl/ctlv3/command/txn_command_test.go
@@ -1,0 +1,91 @@
+// Copyright 2026 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package command
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+)
+
+func TestParseCompare(t *testing.T) {
+	tests := []struct {
+		name       string
+		line       string
+		wantResult pb.Compare_CompareResult
+		wantLease  int64
+		wantErr    bool
+	}{
+		{
+			name:       "issue 20773 regression",
+			line:       `lease("foo1") > "0"`,
+			wantResult: pb.Compare_GREATER,
+			wantLease:  0,
+		},
+		{
+			name:       "hex lease lowercase",
+			line:       `lease("foo1") = "f"`,
+			wantResult: pb.Compare_EQUAL,
+			wantLease:  15,
+		},
+		{
+			name:       "hex lease uppercase",
+			line:       `lease("foo1") = "AF"`,
+			wantResult: pb.Compare_EQUAL,
+			wantLease:  175,
+		},
+		{
+			name:       "long lease id from etcdctl output",
+			line:       `lease("foo1") = "2d8257079fa1bc0c"`,
+			wantResult: pb.Compare_EQUAL,
+			wantLease:  3279279168933706764,
+		},
+		{
+			name:    "invalid hex character",
+			line:    `lease("foo1") > "g"`,
+			wantErr: true,
+		},
+		{
+			name:    "overflow int64 range",
+			line:    `lease("foo1") > "ffffffffffffffff"`,
+			wantErr: true,
+		},
+		{
+			name:    "missing quoted lease value",
+			line:    `lease("foo1") > 10`,
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cmp, err := ParseCompare(tt.line)
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.Equal(t, []byte("foo1"), cmp.Key)
+			require.Equal(t, pb.Compare_LEASE, cmp.Target)
+			require.Equal(t, tt.wantResult, cmp.Result)
+
+			leaseCmp, ok := cmp.TargetUnion.(*pb.Compare_Lease)
+			require.True(t, ok)
+			require.Equal(t, tt.wantLease, leaseCmp.Lease)
+		})
+	}
+}

--- a/etcdctl/ctlv3/command/txn_command_test.go
+++ b/etcdctl/ctlv3/command/txn_command_test.go
@@ -79,11 +79,13 @@ func TestParseCompare(t *testing.T) {
 			}
 
 			require.NoError(t, err)
-			require.Equal(t, []byte("foo1"), cmp.Key)
-			require.Equal(t, pb.Compare_LEASE, cmp.Target)
-			require.Equal(t, tt.wantResult, cmp.Result)
 
-			leaseCmp, ok := cmp.TargetUnion.(*pb.Compare_Lease)
+			pbCmp := cmp.GetCompare()
+			require.Equal(t, []byte("foo1"), pbCmp.Key)
+			require.Equal(t, pb.Compare_LEASE, pbCmp.Target)
+			require.Equal(t, tt.wantResult, pbCmp.Result)
+
+			leaseCmp, ok := pbCmp.TargetUnion.(*pb.Compare_Lease)
 			require.True(t, ok)
 			require.Equal(t, tt.wantLease, leaseCmp.Lease)
 		})

--- a/etcdctl/ctlv3/command/txn_command_test.go
+++ b/etcdctl/ctlv3/command/txn_command_test.go
@@ -18,6 +18,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
+
 	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
 )
 

